torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -704,6 +759,13 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
       at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
       C10D_NCCL_CHECK(
           fn(inputs[i], outputs[i], ncclComms[i]->getNcclComm(), ncclStream));
+      cudaStreamAddCallback(
+          ncclStream,
+          cudaStreamCallback,
+          (void*)std::make_shared<checkFuture>(
+              work, outputs, work_counter, inputs.size(), mutex_)
+              .get(),
```

Comment:
You probably need to use the same `checkFuture` object for all the inputs so that `work_counter` is shared among them.